### PR TITLE
update metrics

### PR DESCRIPTION
**Pull request by mahara13**
this is just raw implementation.
A lot of OverOps types is not  included

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
+/workspace.xml
```
Comment (chook) [inline .idea/.gitignore]: we need to update the main .gitignore with .idea

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+import java.util.*;
+
+public class OverOpsQualityGateStat {
+    public static final String NEW_QG_MARKER = "New";
+    public static final String CRITICAL_QG_MARKER = "Critical";
+    public static final String RESURFACED_QG_MARKER = "Resurfaced";
+    public static final String INCREASING_QG_MARKER = "Increasing";
+
+    public Map<String, RegressionRow> newEventsIds = new HashMap();
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: need diamond <>

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            addNewErrors(rrItem);
+            addIncreasingErrors(rrItem);
+            addCriticalErrors(rrItem);
+            addResurfacedErrors(rrItem);
+        }
+    }
+
+    private void addResurfacedErrors(ReliabilityReport.ReliabilityReportItem rrItem) {
+        for (EventRow row : rrItem.errors) {
+            if((row.labels != null) &&
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: space after if

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                    (row.labels.indexOf("Resurfaced") != -1)) {
+                resurfacedEventsIds.put(row.id, row);
+            }
+        }
+        for (EventRow row : rrItem.failures) {
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: newline before if/for/while/return/switch

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    private void addCriticalErrors(ReliabilityReport.ReliabilityReportItem rrItem) {
+        for (EventRow row : rrItem.failures) {
+            criticalEventsIds.put(row.id, row);
+        }
+    }
+
+    private void addIncreasingErrors(ReliabilityReport.ReliabilityReportItem rrItem) {
+        for (RegressionRow row : rrItem.geIncErrors(true, true)) {
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: it's best to get the hashmap first to a local variable to ease debugging

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        }
+        if (criticalEventsIds.keySet().contains(id)) {
+            result.add(CRITICAL_QG_MARKER);
+        }
+        if (increasingEventsIds.keySet().contains(id)) {
+            result.add(INCREASING_QG_MARKER);
+        }
+
+        if (result.size()==0) {
+            result.add("NO_QUALITY_GATE_IT_SHOULD_BE_VERIFIED");
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: make this a constant as well

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            result.add(RESURFACED_QG_MARKER);
+        }
+        if (criticalEventsIds.keySet().contains(id)) {
+            result.add(CRITICAL_QG_MARKER);
+        }
+        if (increasingEventsIds.keySet().contains(id)) {
+            result.add(INCREASING_QG_MARKER);
+        }
+
+        if (result.size()==0) {
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: isEmpty()

File: src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    private static final String JAVA_LANGUAGE = "java";
+
+    @Override
+    public void register(RegistrarContext registrarContext) {
+        registrarContext.registerClassesForRepository(OverOpsChecks.REPOSITORY_KEY, OverOpsChecks.getChecks(), Collections.EMPTY_LIST);
+    }
+
+    @Override
+    public void define(Context context) {
+        NewRepository repository = context.createRepository(OverOpsChecks.REPOSITORY_KEY, JAVA_LANGUAGE).setName(OVER_OPS_REPOSITORY_NAME);
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java]: please try to break long lines if possible

File: src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        for (Class<? extends JavaCheck> check : OverOpsChecks.getChecks()) {
+            new RulesDefinitionAnnotationLoader().load(repository, new Class[]{check});
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java]: perhaps we need to document or add some debug statement here when verbose debugging is turned on

File: src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            newRule(check, repository);
+        }
+        repository.done();
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java]: newline after }

File: src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    private void addMetadata(NewRule rule, String ruleKey) {
+        OverOpsMetrics.OverOpsMetric ooMetrics = OverOpsMetrics.OverOpsMetric.getByRuleKey(ruleKey);
+        if (ooMetrics != null) {
+            rule.setSeverity(ooMetrics.severity.toString());
+            rule.setName(ooMetrics.ruleTitle);
+            rule.addTags(ooMetrics.ruleTags);
+            rule.setType(ooMetrics.ruleType);
+            rule.setStatus(RuleStatus.READY);
+
+            // TODO find what is proper remediation
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java]: is there a ticket on this?

File: src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    private static String readResource(URL resource) {
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(resource.openStream()))) {
+            return reader.lines().collect(Collectors.joining("\n"));
+        } catch (Exception e) {
+            LOGGER.error("Failed to read: " + resource, e);
+        }
+
+        return null;
+    }
+}
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java]: newlines at end of files

File: src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    @Override
+    public void scanFile(final @Nonnull JavaFileScannerContext ctx) {
+        if (overOpsEventsStatistic == null) {
+            return;
+        }
+
+        context = ctx;
+        file = context.getFile();
+        String filePathJavaStyle = file.getAbsolutePath().replaceAll("/", ".");
+        List<OverOpsEventsStatistic.ClassStat> statForThisFile = overOpsEventsStatistic.getStatistic()
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java]: please do some code cleanup here and perhaps break this to some methods that actually explains what is being done

File: src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        scan(ctx.getTree());
+    }
+
+    public void reportIssue(OverOpsEventsStatistic.LineStat lineStat) {
+        EventResult event = lineStat.event.eventResult;
+        int method_position = event.error_location.original_line_number + 1;
+        long fileCount = getFileCount();
+        boolean isMethodPresent = fileCount >= method_position;
+        method_position = isMethodPresent ? method_position : 1;
+        //log.info(" ");
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java]: cleanup

File: src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsChecks.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.sonar.plugins.java.api.JavaCheck;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+public class OverOpsChecks {
+
+    private static final Collection<Class<? extends JavaCheck>> defaultChecks = new ArrayList<Class<? extends JavaCheck>>();
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsChecks.java]: ArrayList<>();

File: src/main/java/com/overops/plugins/sonar/util/SimpleUrlClient.java [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+import com.takipi.api.core.consts.ApiConstants;
+import com.takipi.api.core.request.intf.ApiPostRequest;
+import com.takipi.api.core.result.intf.ApiResult;
+import com.takipi.api.core.url.UrlClient;
+import com.takipi.common.util.Pair;
+
+import java.net.HttpURLConnection;
+import java.util.Map;
+
+public class SimpleUrlClient extends UrlClient {
```
Comment (chook) [inline src/main/java/com/overops/plugins/sonar/util/SimpleUrlClient.java]: comment about why we need this?

File: .idea/.gitignore (hunk 1 above)
Comment (mahara13) [reply .idea/.gitignore]: Fixed

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java (hunk 2 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: done

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java (hunk 3 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: doen

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java (hunk 4 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: done

File: src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java (hunk 12 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/rules/RuleDefinitionImplementation.java]: done

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java (hunk 5 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: done

File: src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java (hunk 13 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java]: done

File: src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java (hunk 14 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsBaseExceptionCheck.java]: done

File: src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsChecks.java (hunk 15 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/rules/checks/OverOpsChecks.java]: done

File: src/main/java/com/overops/plugins/sonar/util/SimpleUrlClient.java (hunk 16 above)
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/util/SimpleUrlClient.java]: I create this class to be a simple client/request to some API where we put a URL and get a response.
RemoteApiClient - is dedicated to OverOps API calls and has a constructor with package visibility
Therefore I write my own based on UrlClient. Do we need to make it base on BaseApiClient

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
+package com.overops.plugins.sonar.measures;
+
+import com.takipi.api.client.functions.output.EventRow;
+import com.takipi.api.client.functions.output.RegressionRow;
+import com.takipi.api.client.functions.output.ReliabilityReport;
+import com.takipi.api.client.functions.output.ReliabilityReportRow;
+
+import java.util.*;
+
+public class OverOpsQualityGateStat {
```
Comment (daverted) [inline src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: I believe your logic for new, critical, and increasing is correct based on this example: https://github.com/takipi/api-client/blob/ad87f5d912077d3800935808b1a40c5b96dc5707/api-client-util/src/main/java/com/takipi/api/client/functions/sample/OverOpsFunctionsSample.java

I think your logic for resurfaced is also correct, but have not yet been able to verify this.

Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: @daverted Yes, I use the same approach to retrieve

>  new, critical, and increasing

File: src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    public Map<String, EventRow> resurfacedEventsIds = new HashMap<>();
+    public Map<String, EventRow> criticalEventsIds = new HashMap<>();
+    public Map<String, RegressionRow> increasingEventsIds = new HashMap<>();
+
+    public OverOpsQualityGateStat(ReliabilityReport reliabilityReport) {
+        if (reliabilityReport == null || reliabilityReport.items == null) {
+            return;
+        }
+
+        for (Map.Entry<ReliabilityReportRow.Header, ReliabilityReport.ReliabilityReportItem> entry : reliabilityReport.items.entrySet()) {
```
[the remark this answers is not in the document]
Comment (mahara13) [reply src/main/java/com/overops/plugins/sonar/measures/OverOpsQualityGateStat.java]: done

**Comment by daverted** [comment]
@mahara13 do we still need this PR or just #22 ?

**Comment by mahara13** [comment]
> @mahara13 do we still need this PR or just #22 ?

22 contain all changes of this PR

**Comment by daverted** [comment]
superseded by #22

**Pull request closed without merging**